File: 81-search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.cpp
class Solution {
public:
    bool search(vector<int>& nums, int target) {
        int i = 0, j = nums.size()-1;
        while(i <= j){
            int mid = midpoint(i,j);
            if (nums[mid] == target) return true;

            if (nums[i] == nums[mid] && nums[j] == nums[mid]) {
                i++; j--;
                continue;
            }
            if (nums[mid] >= nums[i]){
                if (nums[mid] > target && nums[i] <= target) j = mid - 1;
                else i = mid + 1;
            } else {
                if (target > nums[mid] && target <= nums[j]) i = mid + 1;
                else j = mid - 1;
            }
        }

        return false;
    }
};

File: 81-search-in-rotated-sorted-array-ii/search-in-rotated-sorted-array-ii.go
func search(nums []int, target int) bool {
    left, right := 0, len(nums)-1

    for left <= right {
        mid := left + (right - left)/2

        if nums[mid] == target {
            return true
        } else if nums[left] == nums[right]  {
            left++
        } else if nums[left] <= nums[mid] {
            if target >= nums[left] && target < nums[mid] {
                right = mid - 1
            } else {
                left = mid + 1
            }
        } else {
            if target <= nums[right] && target > nums[mid] {
                left = mid + 1
            } else {
                right = mid - 1
            }
        }
    }

    return false
}